text/0059-nodes-capabilities-discovery.md

```diff
@@ -0,0 +1,142 @@
+
+Implementations must not participate in the main DHT if they don't fulfill the capability yet. For example, a node that is still in the process of warp syncing must not participate in the main DHT. However, assuming that warp syncing doesn't last more than a few seconds, it is acceptable to ignore this requirement in order to avoid complicating implementations too much.
+
+## Drawbacks
+
+None that I can see.
+
+## Testing, Security, and Privacy
+
+*The content of this section is basically the same as the one in RFC 8.*
+
+This mechanism doesn't add or remove any security by itself, as it relies on existing mechanisms.
+
+Due to the way Kademlia works, it would become the responsibility of the 20 Polkadot nodes whose `sha256(peer_id)` is closest to the `key` (described in the explanations section) to store the list of nodes that have specific capabilities.
+Furthermore, when a large number of providers are registered, only the providers closest to the `key` are kept, up to a certain implementation-defined limit.
+
+For this reason, an attacker can abuse this mechanism by randomly generating libp2p PeerIds until they find the 20 entries closest to the `key` representing the target capability. They are then in control of the list of nodes with that capability. While doing this can in no way be actually harmful, it could lead to eclipse attacks.
+
+Because the key changes periodically and isn't predictable, and assuming that the Polkadot DHT is sufficiently large, it is not realistic for an attack like this to be maintained in the long term.
+
+## Performance, Ergonomics, and Compatibility
+
+### Performance
+
+The DHT mechanism generally has a low overhead, especially given that publishing providers is done only every 24 hours.
+
+Doing a Kademlia iterative query then sending a provider record shouldn't take more than around 50 kiB in total of bandwidth for the parachain bootnode.
+
+Assuming 1000 nodes with a specific capability, the 20 Polkadot full nodes corresponding to that capability will each receive a sudden spike of a few megabytes of networking traffic when the `key` rotates. Again, this is relatively negligible. If this becomes a problem, one can add a random delay before a node registers itself to be the provider of the `key` corresponding to `BabeApi_next_epoch`.
+
+Maybe the biggest uncertainty is the traffic that the 20 Polkadot full nodes will receive from light clients that desire knowing the nodes with a capability. If this every becomes a problem, this value of 20 is an arbitrary constant that can be increased for more redundancy.
+
+### Ergonomics
+
+Irrelevant.
+
+### Compatibility
+
+Irrelevant.
+
+## Prior Art and References
+
+Unknown.
+
+## Unresolved Questions
+
+While it fundamentally doesn't change much to this RFC, using `BabeApi_currentEpoch` and `BabeApi_nextEpoch` might be inappropriate. I'm not familiar enough with good practices within the runtime to have an opinion here. Should it be an entirely new pallet?
```

Comment:
elaborate?

Reply:
Maybe we could introduce two runtime functions named for example `NetworkDht_currentRandomness` and `NetworkDht_nextRandomness` that return just the randomness used in the DHT.

From a purely technical point of view, this would be exactly the same as using `BabeApi_currentEpoch` and `BabeApi_nextEpoch`, but it would be more flexible. For example, once we switch to Sassafras, the upgrade would be more seamless than if the usage of Babe is hardcoded.

As the text says, I don't have an opinion here.